From d2fc79b19d733942267ff55df7fd83f352a655a7 Mon Sep 17 00:00:00 2001
From: Vishnu Nair <vishnun@google.com>
Date: Tue, 4 Sep 2018 19:49:57 -0700
Subject: [PATCH 46/58] DO NOT MERGE: Reset launch start times when removing a
 process

When removing an exiting process that is running foreground activities, clear the launch times for
the current windowing mode. When an app process is removed, activities from the process may be
relaunched. In the case of forceStopPackageLocked the activities are finished before any window
is drawn, and the launch time is not cleared. This will be incorrectly used to calculate launch time
for the next launched activity launched in the same windowing mode.

Bug: 80084651
Test: adb shell am start -W com.amazon.mShop.android.shopping/com.amazon.mShop.home.HomeActivity  && sleep 1 && adb shell am force-stop com.amazon.mShop.android.shopping
Change-Id: I2c4f0716c922baa1ad209b4ea1fa7ce366e0e108
(cherry picked from commit 9372683dd2b6ceafee4714770b51e67066e3d8bc)
---
 .../android/server/am/ActivityManagerService.java  | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index f31e614af1d..1e4caf7080a 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -5968,6 +5968,7 @@ public class ActivityManagerService extends IActivityManager.Stub
     private final void handleAppDiedLocked(ProcessRecord app,
             boolean restarting, boolean allowRestart) {
         int pid = app.pid;
+        final boolean clearLaunchStartTime = !restarting && app.removed && app.foregroundActivities;
         boolean kept = cleanUpApplicationRecordLocked(app, restarting, allowRestart, -1,
                 false /*replacingPid*/);
         if (!kept && !restarting) {
@@ -6008,6 +6009,19 @@ public class ActivityManagerService extends IActivityManager.Stub
         } finally {
             mWindowManager.continueSurfaceLayout();
         }
+
+        // Hack for pi
+        // When an app process is removed, activities from the process may be relaunched. In the
+        // case of forceStopPackageLocked the activities are finished before any window is drawn,
+        // and the launch time is not cleared. This will be incorrectly used to calculate launch
+        // time for the next launched activity launched in the same windowing mode.
+        if (clearLaunchStartTime) {
+            final LaunchTimeTracker.Entry entry = mStackSupervisor
+                    .getLaunchTimeTracker().getEntry(mStackSupervisor.getWindowingMode());
+            if (entry != null) {
+                entry.mLaunchStartTime = 0;
+            }
+        }
     }
 
     private final int getLRURecordIndexForAppLocked(IApplicationThread thread) {
-- 
2.17.1

